orchagent/pfcwdorch.cpp

```diff
@@ -0,0 +1,755 @@
+        else
+        {
+            queueEntry.pollTimeLeft = queueEntry.pollTimeLeft == nearestTime ?
+                (queueEntry.handler == nullptr ?
+                 queueEntry.c_detectionTime :
+                 queueEntry.c_restorationTime) :
+                queueEntry.pollTimeLeft - nearestTime;
+        }
+    }
+}
+
+template <typename DropHandler, typename ForwardHandler>
+void PfcWdSwOrch<DropHandler, ForwardHandler>::pfcWatchdogThread(void)
+{
+    DBConnector db(COUNTERS_DB, DBConnector::DEFAULT_UNIXSOCKET, 0);
+
+    // Load script for storm detection
+    string detectScriptName = getStormDetectionCriteria();
+    string detectLuaScript = loadLuaScript(detectScriptName);
+    string detectSha = loadRedisScript(&db, detectLuaScript);
+
+    // Load script for restoration check
+    string restoreLuaScript = loadLuaScript("pfc_restore_check.lua");
+    string restoreSha = loadRedisScript(&db, restoreLuaScript);
+
+    while(m_runPfcWdSwOrchThread)
+    {
+        unique_lock<mutex> lk(m_mtxSleep);
+
+        uint32_t sleepTime = getNearestPollTime();
+
+        m_cvSleep.wait_for(lk, chrono::milliseconds(sleepTime));
+
+        pollQueues(sleepTime, db, detectSha, restoreSha);
+    }
+}
+
+template <typename DropHandler, typename ForwardHandler>
+void PfcWdSwOrch<DropHandler, ForwardHandler>::startWatchdogThread(void)
+{
+    SWSS_LOG_ENTER();
+
+    if (m_runPfcWdSwOrchThread.load())
+    {
+        return;
+    }
+
+    m_runPfcWdSwOrchThread = true;
+
+    m_pfcWatchdogThread = shared_ptr<thread>(
+            new thread(&PfcWdSwOrch::pfcWatchdogThread,
+            this));
+
+    SWSS_LOG_INFO("PFC Watchdog thread started");
+}
+
+template <typename DropHandler, typename ForwardHandler>
+void PfcWdSwOrch<DropHandler, ForwardHandler>::endWatchdogThread(void)
+{
+    SWSS_LOG_ENTER();
+
+    if (!m_runPfcWdSwOrchThread.load())
+    {
+        return;
+    }
+
+    m_runPfcWdSwOrchThread = false;
+
+    m_cvSleep.notify_all();
+
+    if (m_pfcWatchdogThread != nullptr)
+    {
+        SWSS_LOG_INFO("Wait for PFC Watchdog thread to end");
+
+        m_pfcWatchdogThread->join();
+    }
+
+    SWSS_LOG_INFO("PFC Watchdog thread ended");
+}
+
+template <typename DropHandler, typename ForwardHandler>
+PfcDurationWatchdog<DropHandler, ForwardHandler>::PfcDurationWatchdog(
+        DBConnector *db, vector<string> &tableNames):
+    PfcWdSwOrch<DropHandler, ForwardHandler>(db, tableNames)
+{
+    SWSS_LOG_ENTER();
+}
+
+template <typename DropHandler, typename ForwardHandler>
+PfcDurationWatchdog<DropHandler, ForwardHandler>::~PfcDurationWatchdog(void)
+{
+    SWSS_LOG_ENTER();
+}
+
+template <typename DropHandler, typename ForwardHandler>
+vector<sai_port_stat_t> PfcDurationWatchdog<DropHandler, ForwardHandler>::getPortCounterIds(
+        sai_object_id_t queueId)
+{
+    SWSS_LOG_ENTER();
+
+    static const vector<sai_port_stat_t> PfcDurationIdMap =
+    {
+        SAI_PORT_STAT_PFC_0_RX_PAUSE_DURATION,
```

Comment:
other platform does not have this counter, that platform is going to use another counter to do the detection. However, can I specify another counter for another platform?

Reply:
This is PFC duration based WD. Other platform should inherit from base class and register their own counters.